nixos/modules/system/boot/systemd.nix

```diff
@@ -695,6 +697,20 @@ in
         minsize = "1M";
       };
     };
+
+    # make sure /run/next-system isn't garbage collected before the next boot
+    systemd.tmpfiles.rules = [ "L+ /nix/var/nix/gcroots/next-system - - - - /run/next-system" ];
+
+    # if /run/next-system exists, activate it before  rebooting
+    systemd.services.systemd-soft-reboot-activate-next-system = {
+      wantedBy = [ "soft-reboot.target" ];
+      before = [ "systemd-soft-reboot.service" ];
+      unitConfig = {
+        DefaultDependencies = false;
+        ConditionPathExists = "/run/next-system/activate";
+      };
+      serviceConfig.ExecStart = "/run/next-system/activate";
```

Comment:
Should this fall back to `/run/current-system/activate` if `/run/next-system/activate` doesn't exist?

Reply:
Same argument as not creating `next-system` on `switch`; the current system is already activated so it shouldn't be necessary.